pkg/compiler/types.go

```diff
@@ -152,8 +155,13 @@ var typeArray = &typeDesc{
 	Gen: func(comp *compiler, t *ast.Type, args []*ast.Type, base prog.IntTypeCommon) prog.Type {
 		elemType := comp.genType(args[0], "", base.ArgDir, false)
 		kind, begin, end := prog.ArrayRandLen, uint64(0), uint64(0)
+		literal := ""
 		if len(args) > 1 {
-			kind, begin, end = prog.ArrayRangeLen, args[1].Value, args[1].Value2
+			if args[1].Ident == "len" {
+				literal = args[1].Args[0].Ident
```

Comment:
Added `ArrayLenTarget`

Reply:
BTW, how do we go about generating arrays of the `ArrayLenTarget` kind? I saw that `LenType` arguments are fixed up in `assignSizes` as another pass separate from argument generation.

Should we somehow make sure that generateArgs populates the target array first, so that the sizes of those arrays can be found by `ArrayLenTarget` arrays?

Reply:
I don't have precise answer.
I think we should restrict len to refer to a _preceding_ array only. Then memorize array sizes in a map as we generate fields, and then when we generate a parallel array consult that map. Something along these lines.

Reply:
Ok, I got the idea. I will ping you once I have a working version.